### PR TITLE
Grid returns ranges for x, y, z

### PR DESCRIPTION
**Pull request by navidcy**

### CODE REVIEW

**Comment by navidcy** [comment]
no, problem was in 
```julia
c0 = c0func.(g.x)
```
which was creating a CPU array `c0` regardless of `dev`.  Changed that to
```julia
  x = ArrayType(dev)(g.x)
  c0 = c0func.(x)
```

Testing on bruno before I push...

File: src/domains.jl [hunk 1]
```diff
@@ -31,19 +31,20 @@ Constructs a OneDGrid object with size `Lx`, resolution `nx`, and leftmost
 position `x0`. FFT plans are generated for `nthreads` CPUs using
 FFTW flag `effort`.
 """
-struct OneDGrid{T<:AbstractFloat, Ta<:AbstractArray, Tfft, Trfft} <: AbstractGrid{T, Ta}
+struct OneDGrid{T<:AbstractFloat, Tk, Tx, Tfft, Trfft} <: AbstractGrid{T, Tk}
```
Comment (glwagner) [inline src/domains.jl:34]: Where is the second type parameter in `AbstractGrid` used? What's its purpose?

**Comment by glwagner** [comment]
I think the solution I suggested is more user-friendly because it does not require users to write `ArrayType` in their scripts. I think we should support patterns that allow users to pass CPU arrays to `set!`. This example is supposed to serve as an example for future packages. What do you think?

> no, problem was in
> 
> ```julia
> c0 = c0func.(g.x)
> ```
> 
> which was creating a CPU array `c0` regardless of `dev`. Changed that to
> 
> ```julia
>   x = ArrayType(dev)(g.x)
>   c0 = c0func.(x)
> ```
> 
> Testing on bruno before I push...

File: src/domains.jl (hunk 1 above)
Comment (navidcy) [reply src/domains.jl:34]: Some functions use this type of array (`Array` or `CuArray`), e.g.,
https://github.com/FourierFlows/FourierFlows.jl/blob/999b1ef28a3d1010d7bf9ae31c48a65fd708f321/src/domains.jl#L277-L293

Comment (navidcy) [reply src/domains.jl:34]: It's also used a lot in GeophysicalFlows.jl

**Comment by navidcy** [comment]
@glwagner, OK I agree!
(I had misunderstood your previous comment.)

However the suggested solution you have doesn't work. If you try to copy values of an Array to a CuArray it's not feasible...

```
julia> b = Array(rand(3,1)); b2 = 0*deepcopy(b);

julia> a = CuArray(rand(3,1)); a2 = 0*deepcopy(a);

julia> a
3×1 CuArray{Float64,2,Nothing}:
 0.7800711678587244
 0.7696059213580613
 0.1258728203788304

julia> b
3×1 Array{Float64,2}:
 0.9767737074659868
 0.5158999408061726
 0.5504044087481377

julia> a2
3×1 CuArray{Float64,2,Nothing}:
 0.0
 0.0
 0.0

julia> b2
3×1 Array{Float64,2}:
 0.0
 0.0
 0.0

julia> b2 .= a
3×1 Array{Float64,2}:
 0.7800711678587244
 0.7696059213580613
 0.1258728203788304

julia> a2 .= b
ERROR: GPU compilation of kernel broadcast(CuArrays.CuKernelContext, CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}, which is not isbits:
  .args is of type Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}} which is not isbits.
    .1 is of type Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}} which is not isbits.
      .x is of type Array{Float64,2} which is not isbits.

Passing non-isbits types is only allowed if they they are unused by the kernel.

Stacktrace:
 [1] check_invocation(::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDAnative.CUDACompilerParams}, ::LLVM.Function) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/validation.jl:75
 [2] macro expansion at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/driver.jl:184 [inlined]
 [3] macro expansion at /home/navid/.julia/packages/TimerOutputs/NvIUx/src/TimerOutput.jl:245 [inlined]
 [4] #codegen#67(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(GPUCompiler.codegen), ::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDAnative.CUDACompilerParams}) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/driver.jl:183
 [5] #codegen at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/driver.jl:0 [inlined]
 [6] #compile#65(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(GPUCompiler.compile), ::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDAnative.CUDACompilerParams}) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/driver.jl:36
 [7] #compile at ./none:0 [inlined]
 [8] #_cufunction#122(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative._cufunction), ::GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}) at /home/navid/.julia/packages/CUDAnative/e0IdN/src/execution.jl:308
 [9] _cufunction at /home/navid/.julia/packages/CUDAnative/e0IdN/src/execution.jl:302 [inlined]
 [10] #77 at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/cache.jl:21 [inlined]
 [11] get!(::GPUCompiler.var"#77#78"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},typeof(CUDAnative._cufunction),GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}}, ::Dict{UInt64,Any}, ::UInt64) at ./dict.jl:452
 [12] macro expansion at ./lock.jl:173 [inlined]
 [13] #check_cache#76(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(GPUCompiler.check_cache), ::typeof(CUDAnative._cufunction), ::GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}, ::UInt64) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/cache.jl:19
 [14] (::GPUCompiler.var"#kw##check_cache")(::NamedTuple{(),Tuple{}}, ::typeof(GPUCompiler.check_cache), ::Function, ::GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}, ::UInt64) at ./none:0
 [15] hash_64_64 at ./int.jl:53 [inlined]
 [16] hash_uint64 at ./hashing.jl:62 [inlined]
 [17] hx at ./float.jl:567 [inlined]
 [18] hash at ./float.jl:570 [inlined]
 [19] #cached_compilation#79(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(GPUCompiler.cached_compilation), ::typeof(CUDAnative._cufunction), ::GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}, ::UInt64) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/cache.jl:0
 [20] cached_compilation(::Function, ::GPUCompiler.FunctionSpec{GPUArrays.var"#28#29",Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Float64,2,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}}, ::UInt64) at /home/navid/.julia/packages/GPUCompiler/bwcs0/src/cache.jl:37
 [21] #cufunction#121(::String, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::Function, ::Type) at /home/navid/.julia/packages/CUDAnative/e0IdN/src/execution.jl:296
 [22] #cufunction at ./none:0 [inlined]
 [23] macro expansion at /home/navid/.julia/packages/CUDAnative/e0IdN/src/execution.jl:108 [inlined]
 [24] macro expansion at ./gcutils.jl:91 [inlined]
 [25] macro expansion at /home/navid/.julia/packages/CUDAnative/e0IdN/src/execution.jl:105 [inlined]
 [26] #gpu_call#49(::String, ::typeof(GPUArrays.gpu_call), ::CuArrays.CuArrayBackend, ::Function, ::Tuple{CuArray{Float64,2,Nothing},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,2},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}}, ::Int64) at /home/navid/.julia/packages/CuArrays/l0gXB/src/gpuarrays.jl:32
 [27] #gpu_call at /home/navid/.julia/packages/GPUArrays/OXvxB/src/device/execution.jl:0 [inlined]
 [28] #gpu_call#1 at /home/navid/.julia/packages/GPUArrays/OXvxB/src/device/execution.jl:61 [inlined]
 [29] #gpu_call at ./none:0 [inlined]
 [30] copyto! at /home/navid/.julia/packages/GPUArrays/OXvxB/src/host/broadcast.jl:63 [inlined]
 [31] copyto! at ./broadcast.jl:863 [inlined]
 [32] materialize!(::CuArray{Float64,2,Nothing}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2},Nothing,typeof(identity),Tuple{Array{Float64,2}}}) at ./broadcast.jl:822
 [33] top-level scope at REPL[97]:1
```

I  ended up using the parameter of Problem type to convert the input of `set_c!` to the appropriate type
```julia
set_c!(prob::FourierFlows.Problem{T, A}, c) where {T, A}
  prob.vars.c .= A(c) # this converts c to the ArrayType used in prob.vars.c (e.g., convert to CuArray if user gives c as Arrray)
...
```

Let me know if you have a more elegant solution in mind. Perhaps we could consider clearing up the `Problem` struct. One idea is that `dev` should be a parameter of problem. Then every function with `problem` as its argument would know if this problem lives on CPU or GPU.

**Pull request merged**